microceph/ceph/operations.go

```diff
@@ -0,0 +1,284 @@
+	return "Run `ceph osd set noout`."
+}
+
+// AssertNooutFlagSetOps is an operation to assert noout has been set for the ceph cluster.
+type AssertNooutFlagSetOps struct {
+	ClusterOps
+}
+
+// Run asserts noout has been set for the ceph cluster.
+func (o *AssertNooutFlagSetOps) Run(name string) error {
+	set, err := isOsdNooutSet()
+	if err != nil {
+		return err
+	}
+	if !set {
+		return fmt.Errorf("osd has 'noout' flag unset.")
+	}
+	logger.Info("osd has 'noout' flag set.")
+	return nil
+}
+
+// DryRun prints out the action plan.
+func (o *AssertNooutFlagSetOps) DryRun(name string) string {
+	return "Assert osd has 'noout' flag set."
+}
+
+// AssertNooutFlagUnsetOps is an operation to assert noout has been unset for the ceph cluster.
+type AssertNooutFlagUnsetOps struct {
+	ClusterOps
+}
+
+// Run asserts noout has been unset for the ceph cluster.
+func (o *AssertNooutFlagUnsetOps) Run(name string) error {
+	set, err := isOsdNooutSet()
+	if err != nil {
+		return err
+	}
+	if set {
+		return fmt.Errorf("osd has 'noout' flag set.")
+	}
+	logger.Info("osd has 'noout' flag unset.")
+	return nil
+}
+
+// DryRun prints out the action plan.
+func (o *AssertNooutFlagUnsetOps) DryRun(name string) string {
+	return "Assert osd has 'noout' flag unset."
+}
+
+// StopOsdOps is an operation to stop osd service for a node.
+type StopOsdOps struct {
+	ClusterOps
+}
+
+// Run stops the osd service for a node.
+func (o *StopOsdOps) Run(name string) error {
+	err := o.CephClient.PutOsds(o.ClusterClient, false, name)
+	if err != nil {
+		logger.Errorf("unable to stop osd service in node '%s': %v", name, err)
+		return err
+	}
+	logger.Infof("stopped osd service in node '%s'.", name)
+	return nil
+}
+
+// DryRun prints out the action plan.
+func (o *StopOsdOps) DryRun(name string) string {
+	return fmt.Sprintf("Stop osd service in node '%s'.", name)
+}
+
+// StartOsdOps is an operation to start osd service for a node.
+type StartOsdOps struct {
+	ClusterOps
+}
+
+// Run starts the osd service for a node.
+func (o *StartOsdOps) Run(name string) error {
+	err := o.CephClient.PutOsds(o.ClusterClient, true, name)
+	if err != nil {
+		logger.Errorf("unable to start osd service in node '%s': %v", name, err)
+		return err
+	}
+	logger.Infof("started osd service in node '%s'", name)
+	return nil
+}
+
+// DryRun prints out the action plan.
+func (o *StartOsdOps) DryRun(name string) string {
+	return fmt.Sprintf("Start osd service in node '%s'.", name)
+}
+
+// UnsetNooutOps is an operation to unset noout for the ceph cluster.
+type UnsetNooutOps struct {
+	ClusterOps
+}
+
+// Run `ceph osd unset noout` for the ceph cluster.
+func (o *UnsetNooutOps) Run(name string) error {
+	err := setOsdNooutFlag(false)
+	if err != nil {
+		return err
+	}
+	logger.Info("unset osd noout.")
+	return nil
+}
+
+// DryRun prints out the action plan.
+func (o *UnsetNooutOps) DryRun(name string) string {
+	return "Run `ceph osd unset noout`."
+}
```

Comment:
There's a little bit of abstraction leaking between Run and DryRun here: Run uses the setOsdNooutFlag() function, but DryRun prints the lower level ceph command used. Just an observation. Not sure how nitpicky this is, or the best way to go - perhaps inlining the setOsdNooutFlag function to remove the extra layer of indirection?

Reply:
The `DryRun` function should give you the actual action performed, not the string representation of the function